parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecPool.java

```diff
@@ -0,0 +1,239 @@
+        if (decom != null) {
+          dePools.put(decom.getClass(), decompressorPool);
+          decompressorPool.returnObject(decom);
+        } else {
+          if (Log.WARN) {
+            LOG.warn("Unable to find decompressor for codec " + codec.getClass().getName());
+          }
+        }
+
+        if (supportDirectDecompressor) {
+          directDecompressorPool = new GenericObjectPool(
+              new BasePoolableObjectFactory() {
+            public Object makeObject() throws Exception {
+              return CREATE_DIRECT_DECOMPRESSOR_METHOD.invoke(DIRECT_DECOMPRESSION_CODEC_CLASS);
+            }
+          }, Integer.MAX_VALUE);
+
+          Object ddecom = directDecompressorPool.borrowObject();
+          if (ddecom != null) {
+            directDePools.put(ddecom.getClass(), directDecompressorPool);
+            directDecompressorPool.returnObject(ddecom);
+
+          } else {
+            supportDirectDecompressor = false;
+            if (Log.WARN) {
+              LOG.warn("Unable to find direct decompressor for codec {}" + codec.getClass().getName());
+            }
+          }
+
+        } else {
+          directDecompressorPool = null;
+        }
+
+        this.supportDirectDecompressor = supportDirectDecompressor;
+      } catch (Exception e) {
+        throw new ParquetRuntimeException("Error creating compression codec pool.") {
+          // This exception will not likely be recoverable
+        };
+      }
+    }
+
+    public Object borrowDirectDecompressor(){
+      Preconditions.checkArgument(supportDirectDecompressor, "Tried to get a direct Decompressor from a non-direct codec.");
+      try {
+        return directDecompressorPool.borrowObject();
+      } catch (Exception e) {
+        throw new ParquetCompressionCodecException(e);
+      }
+    }
+
+    public boolean supportsDirectDecompression() {
+      return supportDirectDecompressor;
+    }
+
+    public Decompressor borrowDecompressor(){
+      try {
+        return (Decompressor) decompressorPool.borrowObject();
+      } catch (Exception e) {
+        throw new ParquetCompressionCodecException(e);
+      }
+    }
+
+    public Compressor borrowCompressor(){
+      try {
+        return (Compressor) compressorPool.borrowObject();
+      } catch (Exception e) {
+        throw new ParquetCompressionCodecException(e);
+      }
+    }
+  }
+
+  public CodecPool codec(CompressionCodec codec){
+    CodecPool pools = codecs.get(codec);
+    if(pools == null){
+      synchronized(this){
+        pools = codecs.get(codec);
+        if(pools == null){
+          pools = new CodecPool(codec);
+          codecs.put(codec, pools);
+        }
+      }
+    }
+    return pools;
+  }
+
+  private void returnToPool(Object obj, Map<Class<?>, GenericObjectPool> pools) {
+    try {
+      GenericObjectPool pool = pools.get(obj.getClass());
+      if (pool == null) {
+        throw new IllegalStateException("Received unexpected decompressor.");
```

Comment:
Added the class name of the object passed and tried to make the message a little clearer.